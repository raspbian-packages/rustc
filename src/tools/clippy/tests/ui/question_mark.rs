fn some_func(a: Option<u32>) -> Option<u32> {
	if a.is_none() {
		return None
	}

	a
}

pub enum SeemsOption<T> {
    Some(T),
    None
}

impl<T> SeemsOption<T> {
    pub fn is_none(&self) -> bool {
        match *self {
            SeemsOption::None => true,
            SeemsOption::Some(_) => false,
        }
    }
}

fn returns_something_similar_to_option(a: SeemsOption<u32>) -> SeemsOption<u32> {
    if a.is_none() {
        return SeemsOption::None;
    }

    a
}

pub struct SomeStruct {
	pub opt: Option<u32>,
}

impl SomeStruct {
	pub fn func(&self) -> Option<u32> {
		if (self.opt).is_none() {
			return None;
		}

		self.opt
	}
}

fn main() {
	some_func(Some(42));
	some_func(None);

	let some_struct = SomeStruct { opt: Some(54) };
	some_struct.func();

    let so = SeemsOption::Some(45);
    returns_something_similar_to_option(so);
}
