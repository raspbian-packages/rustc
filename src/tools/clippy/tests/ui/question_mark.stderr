error: this block may be rewritten with the `?` operator
 --> $DIR/question_mark.rs:2:2
  |
2 |       if a.is_none() {
  |  _____^
3 | |         return None
4 | |     }
  | |_____^ help: replace_it_with: `a?;`
  |
  = note: `-D question-mark` implied by `-D warnings`

error: this block may be rewritten with the `?` operator
  --> $DIR/question_mark.rs:37:3
   |
37 |           if (self.opt).is_none() {
   |  _________^
38 | |             return None;
39 | |         }
   | |_________^ help: replace_it_with: `(self.opt)?;`

error: aborting due to 2 previous errors

